File: strategy/typescript/glados/test/move/victory.ts
import { FriendlyRobot } from "base/robot";
import { Position, Vector } from "base/vector";
import * as vis from "base/vis";
import * as World from "base/world";
let G = World.Geometry;

import { MessageBox } from "glados/control/messaging";
import { Assignment, Move, MoveParameters } from "glados/group/move/base";
import { MoveToPos } from "glados/task/shared/movetopos";
import { Victory as VictoryTask } from "glados/task/test/victory";

export class Victory extends Move {

	public static readonly MIN_ROBOTS = 3;
	public static readonly MAX_ROBOTS = 12;
	public static readonly ALLOW_EXTRA_ATTACKERS = false;

	_state: string;


	constructor(robots: FriendlyRobot[], messaging: MessageBox) {
		super(robots, messaging);
		this._state = "init";
	}

	static canStart() {
		return true;
	}

	_canContinue() {
		return true;
	}

	_updateTasks(): MoveParameters {
		let taskAssignments: Map<FriendlyRobot, Assignment> = new Map<FriendlyRobot, Assignment>();

		let nRobots = this._robots.length;
		let radius = (G.FieldHeightHalf - G.DefenseHeight) / 2;
		let center: Position = new Vector(0, -radius - 0.75);
		radius = radius - 0.5;
		vis.addCircle("test", center, 0.05, vis.colors.yellow, true);
		let angleStep = 2 * Math.PI / nRobots;

		if (this._state === "init") {
			for (let i = 0; i < this._robots.length; i++) {
				let angle = i * angleStep;
				let moveLine = Vector.fromPolar(angle, radius / 2);
				let pos = center - new Vector(0, -radius / 2) + moveLine;
				taskAssignments[this._robots[i]] = { class: MoveToPos, params: [pos]};
				if (this._robots[i].pos.distanceTo(pos) > 0.1) {
					this._state = "circle";
				}
			}
		} else if (this._state === "circle") {
			for (let i = 0; i < this._robots.length; i++) {
				let angle = (i - 1) * angleStep;
				taskAssignments[this._robots[i]] = { class: VictoryTask, params: [center, 0, angle, radius]};
			}
		}
		let mainAttacker: FriendlyRobot | undefined = undefined;
		return {
			assignments: taskAssignments,
			mainAttacker: mainAttacker
		};
	}
}

File: strategy/lua/marvin/test/move/victory.lua
--[[***********************************************************************
*   Copyright 2026 Robotics Erlangen e.V.                                 *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, or     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
*************************************************************************]]

local Victory = Class("Group.Move.Victory", require "group/move/base")

local World = require "../base/world"
local G = World.Geometry

local MoveToPos = require "task/shared/movetopos"
local VictoryTask = require "task/test/victory"

local vis = require "../base/vis"

Victory.MIN_ROBOTS = 3
Victory.MAX_ROBOTS = 12

function Victory.canStart() -- TODO
	return true
end

function Victory:_init()
	self._state = "init"
end

function Victory:_canContinue() -- TODO
	return true
end

function Victory:_updateTasks()
	local taskAssignments = {}

	local nRobots = #self._robots
	-- TODO: radius sinnvoller
	local radius = (G.FieldHeightHalf - G.DefenseRadius) / 2
	local center = Vector(0, -radius - 0.75)
	radius = radius - 0.5
	vis.addCircle("test", center, 0.05, vis.colors.yellow, true)
	local angleStep = 2 * math.pi / nRobots

	if self._state == "init" then -- todo startposition fixen
		for i, _ in ipairs(self._robots) do
			local angle = i * angleStep
			local moveLine = Vector.fromAngle(angle):setLength(radius/2)
			local pos = center - Vector(0, -radius/2) + moveLine
			taskAssignments[self._robots[i]] = { class = MoveToPos, params = {pos}}
			if self._robots[i].pos:distanceTo(pos) > 0.1 then
				self._state = "circle"
			end
		end
	elseif self._state == "circle" then
		for i, _ in ipairs(self._robots) do
			local angle = (i-1) * angleStep
			taskAssignments[self._robots[i]] = { class = VictoryTask, params = {center, 0, angle, radius}}
		end
	end

	return taskAssignments
end
return Victory
